### PR TITLE
feat: add keyboard shortcuts to copy and paste quests

### PR DESCRIPTION
**Pull request by adamico**
Ctrl + C copies the selected quest and a toast is displayed with the quest title Ctrl + V pastes the quest
Ctrl + Shift + V pastes without dependencies
Ctrl + Alt + V pastes the quest link

copy-paste-with-keys

Notes for things to tackle in separate PRs
1. when doing Ctrl + C with multiple quests selected an error toast is displayed
2. when pasting a quest it would be useful not having the pasted quest opening its panel

closes https://github.com/FTBTeam/FTB-Mods-Issues/issues/973

### CODE REVIEW

File: common/src/main/java/dev/ftb/mods/ftbquests/client/gui/quests/QuestScreen.java [hunk 1]
```diff
@@ -406,6 +409,63 @@ public boolean keyPressed(Key key) {
 					return moveSelectedQuests(step, 0D);
 				}
 			}
+			switch (key.keyCode) {
+				case GLFW.GLFW_KEY_C -> {
+					Collection<Quest> quests = getSelectedQuests();
```
Comment (desht) [inline common/src/main/java/dev/ftb/mods/ftbquests/client/gui/quests/QuestScreen.java:414]: Would like to see this split out into its own method for tidiness.  It's too much code for a switch case.

File: common/src/main/java/dev/ftb/mods/ftbquests/client/gui/quests/QuestScreen.java [hunk 2]
```diff
@@ -406,6 +409,63 @@ public boolean keyPressed(Key key) {
...
+						toastDesc = Component.literal(qTitle);
+
+						setClipboardString(qId);
+						Minecraft.getInstance().getToasts().addToast(new CustomToast(Component.translatable(toastTitleKey), icon, toastDesc));
+					}
+
+					return true;
+				}
+				case GLFW.GLFW_KEY_V -> {
+					double minSize = selectedObjects.stream()
```
Comment (desht) [inline common/src/main/java/dev/ftb/mods/ftbquests/client/gui/quests/QuestScreen.java:437]: See my comment on line 414

File: common/src/main/java/dev/ftb/mods/ftbquests/net/CopyQuestMessage.java [hunk 3]
```diff
@@ -36,6 +36,9 @@ public CopyQuestMessage(FriendlyByteBuf buf) {
         this.qy = buf.readDouble();
         this.copyDeps = buf.readBoolean();
     }
+    public void setCopyDeps(boolean b) {
```
Comment (desht) [inline common/src/main/java/dev/ftb/mods/ftbquests/net/CopyQuestMessage.java:39]: Unnecessary method

**Comment by desht** [comment]
Looks good in principle but please see my comments on changes that are needed.  The same changes will also need to be applied to the 1.19 PR (having said that, I don't mind doing a cherry-pick of this into 1.19 once it's merged, so you can close the 1.19 PR if you prefer).

File: common/src/main/java/dev/ftb/mods/ftbquests/client/gui/quests/QuestScreen.java (hunk 1 above)
Comment (adamico) [reply common/src/main/java/dev/ftb/mods/ftbquests/client/gui/quests/QuestScreen.java:414]: yup, I'll refactor into separate methods to use with right click + copy or paste

File: common/src/main/java/dev/ftb/mods/ftbquests/client/gui/quests/QuestScreen.java [hunk 4]
```diff
@@ -406,6 +409,63 @@ public boolean keyPressed(Key key) {
...
+							long questId = Long.valueOf(clip, 16);
+							QuestObject qo = file.get(questId);
+							if (qo instanceof Quest quest) {
+								CopyQuestMessage cQMessage = new CopyQuestMessage(quest, chapter, qx, qy, false);
+								if (key.modifiers.alt()) {
+									QuestLink link = new QuestLink(0L, chapter, questId);
+									link.setPosition(qx, qy);
+									new CreateObjectMessage(link, new CompoundTag()).sendToServer();
+								} else {
+                                    cQMessage.setCopyDeps(!key.modifiers.shift());
```
[the remark this answers is not in the document]
Comment (adamico) [reply common/src/main/java/dev/ftb/mods/ftbquests/client/gui/quests/QuestScreen.java:457]: yeah, I'll fix this

**Comment by adamico** [comment]
ok to close the 1.19 pr and cherry-pick the 1.20 changes

I'll make the changes and submit new commits to the PR

**Comment by adamico** [comment]
> Having newly created quests always open their panel is something I intend to address before the next FTB Quests release (there's also an issue of quest creation currently opening the panel for other players on the server in a SMP setting, which I've already fixed in dev).

Ok for me.

**Comment by adamico** [comment]
I'm done @desht

**Comment by desht** [comment]
Thanks, should hopefully get to this in a day or two (haven't forgotten, this is definitely useful functionality to get merged!)

**Pull request merged**